File: src/main/java/com/leet/medium/MinimumMovesToEqualArrayElementsII.exs
defmodule Solution do
  @spec min_moves2(nums :: [integer]) :: integer
  def min_moves2([num]) do
    0
  end

  @spec min_moves2(nums :: [integer]) :: integer
  def min_moves2([]) do
    0
  end

  @spec min_moves2(nums :: [integer]) :: integer
  def min_moves2(nums) do
    sorted = Enum.sort(nums)
    size = length(nums)

    median =
      case rem(size, 2) == 0 do
        false ->
          Enum.at(sorted, div(size, 2))

        true ->
          mid_idx = div(size, 2)
          div(Enum.at(sorted, mid_idx) + Enum.at(sorted, mid_idx - 1), 2)
      end

    Enum.sum(Enum.map(sorted, fn e -> abs(e - median) end))
  end
end

File: src/main/java/com/leet/medium/MinimumMovesToEqualArrayElementsII.go
package main

import "sort"

func minMoves2(nums []int) int {
	sort.Ints(nums)
	median := nums[len(nums)/2]
	if len(nums)%2 != 1 {
		median = (median + nums[len(nums)/2-1]) / 2
	}
	ans := 0
	for _, item := range nums {
		val := item - median
		if val < 0 {
			ans += -val
		} else {
			ans += val
		}
	}
	return ans
}
